### PR TITLE
jewel: rgw: libcurl & ssl fixes 

### PR DESCRIPTION
**Pull request by theanalyst**
Fixes:
http://tracker.ceph.com/issues/23243
http://tracker.ceph.com/issues/23240 
http://tracker.ceph.com/issues/23244

Commit 788bbd9be22d33d460d3e0bcf76cbf2e55a822b2 is jewel only and ports the cmake like changes to autotools. Combining all the fixes into one PR as merging them in various orders would cause different sorts of merge conflicts when trying to build an integration branch for teuthology.

### CODE REVIEW

File: src/rgw/rgw_http_client.cc [hunk 1]
```diff
@@ -75,6 +75,135 @@ struct rgw_http_req_data : public RefCountedObject {
   }
 };
 
+struct RGWCurlHandle {
+  int uses;
+  mono_time lastuse;
```
Comment (chardan) [inline src/rgw/rgw_http_client.cc]: Did you mean to leave this uninitialized?

File: src/rgw/rgw_http_client.cc [hunk 2]
```diff
@@ -75,6 +75,135 @@ struct rgw_http_req_data : public RefCountedObject {
...
+  CURL* h;
+
+  RGWCurlHandle(CURL* h) : uses(0), h(h) {};
+  CURL* operator*() {
+    return this->h;
+  }
+};
+
+#define MAXIDLE 5
```
Comment (chardan) [inline src/rgw/rgw_http_client.cc]: Nit: prefer constexpr to macro constants.

File: src/rgw/rgw_http_client.cc [hunk 3]
```diff
@@ -75,6 +75,135 @@ struct rgw_http_req_data : public RefCountedObject {
...
+class RGWCurlHandles : public Thread {
+public:
+  Mutex cleaner_lock;
```
Comment (chardan) [inline src/rgw/rgw_http_client.cc]: Nothing wrong with this, but std::mutex is also worth considering (and lighter-weight than Ceph::Mutex).

File: src/rgw/rgw_http_client.cc [hunk 4]
```diff
@@ -75,6 +75,135 @@ struct rgw_http_req_data : public RefCountedObject {
...
+    if (!saved_curl.empty()) {
+      curl = *saved_curl.begin();
+      saved_curl.erase(saved_curl.begin());
+    }
+  }
+  if (curl) {
+  } else if ((h = curl_easy_init())) {
+    curl = new RGWCurlHandle{h};
+  } else {
+    // curl = 0;
```
Comment (chardan) [inline src/rgw/rgw_http_client.cc]: ?

**Comment by smithfarm** [comment]
@chardan Please note that this is a backport PR, so the commits are cherry-picked from master with any changes kept to a bare minimum and described in the "Conflicts" section of each commit.

**Comment by smithfarm** [comment]
This passed two rgw runs at http://tracker.ceph.com/issues/21742#note-17

OK to merge?

**Pull request merged**